### PR TITLE
[v7r2] Support hypothesis 6 (#4892)

### LINKED ISSUE #4892: Tests fail with hypothesis 6
From the [release notes](https://hypothesis.readthedocs.io/en/latest/changes.html#v6-0-0):

> Using function-scoped pytest fixtures is now a health-check error, instead of a warning.

this results in the error:

> hypothesis.errors.FailedHealthCheck: src/DIRAC/Core/Security/test/Test_X509Chain.py::test_delegation[M2_X509Request-M2_X509Chain] uses the 'get_X509Request' fixture, which is reset between function calls but not between test cases generated by `@given(...)`.  You can change it to a module- or session-scoped fixture if it is safe to reuse; if not we recommend using a context manager inside your test function.  See https://docs.pytest.org/en/latest/fixture.html#sharing-test-data for details on fixture scope.
>
> See https://hypothesis.readthedocs.io/en/latest/healthchecks.html for more information about this. If you want to disable just this health check, add HealthCheck.function_scoped_fixture to the suppress_health_check settings for this test.

Example log: https://github.com/DIRACGrid/DIRAC/runs/1694028407?check_suite_focus=true

### PR DESCRIPTION
**Pull request by chrisburr**
Closes #4892 by uses a context manager instead of a fixture.

### CODE REVIEW

File: src/DIRAC/Core/Security/test/Test_X509Chain.py [hunk 1]
```diff
@@ -49,53 +49,43 @@
...
 
 
-@fixture(scope="function", params=X509CHAINTYPES)
-def get_proxy(request):
-  """ Fixture to return either the proxy string.
-      It also 'de-import' DIRAC before and after
-  """
-  # Clean before
+@contextlib.contextmanager
+def get_proxy(x509ChainClassName, certFile, lifetime=3600, deimport=True, **kwargs):
```
Comment (chaen) [inline src/DIRAC/Core/Security/test/Test_X509Chain.py:53]: add the new parameter docstring

File: src/DIRAC/Core/Security/test/Test_X509Chain.py [hunk 2]
```diff
@@ -515,9 +500,11 @@ def test_getIssuerCert(get_proxy):
   #                                               lifetime=chainLifeTime,
   #                                               diracGroup=diracGroup,
   #                                               rfc = rfcIfPossible)
+@parametrize("x509ClassName", X509REQUESTTYPES)
+@parametrize('x509ChainClassName', X509CHAINTYPES)
 @settings(max_examples=200)
 @given(diracGroup=text(ascii_letters + '-', min_size=1), lifetime=integers(min_value=1, max_value=TWENTY_YEARS_IN_SEC))
-def test_delegation(get_X509Request, get_proxy, diracGroup, lifetime):
+def test_delegation(x509ClassName, x509ChainClassName, diracGroup, lifetime):
```
Comment (chaen) [inline src/DIRAC/Core/Security/test/Test_X509Chain.py:507]: You can change the comment saying that the name is important :-)

File: src/DIRAC/Core/Security/test/Test_X509Chain.py [hunk 3]
```diff
@@ -529,51 +516,50 @@ def test_delegation(get_X509Request, get_proxy, diracGroup, lifetime):
...
 
-  # The client side signs the request
+    #######################################################
 
-  proxyChain = get_proxy(USERCERT, diracGroup=diracGroup)
+    # The client side signs the request
 
-  # The proxy will contain a "bullshit private key"
-  res = proxyChain.generateChainFromRequestString(reqStr, lifetime=lifetime)
+    with get_proxy(x509ChainClassName, USERCERT, diracGroup=diracGroup, deimport=False) as proxyChain:
```
Comment (chaen) [inline src/DIRAC/Core/Security/test/Test_X509Chain.py:530]: Maybe add a comment here to explain why `deimport` is False (copy paste what was in the previous incarnation of `get_proxy` ? )

**Comment by chrisburr** [comment]
@chaen Thanks for the quick review, I've applied your suggestions.

I also found the DISET tests have similar issues but converting them to use a context manager wasn't obvious so I've disabled this health check exclusively for those tests.

File: src/DIRAC/Core/Utilities/test/Test_Encode.py [hunk 4]
```diff
@@ -27,6 +27,11 @@
 from pytest import mark, approx, raises, fixture, skip
 parametrize = mark.parametrize
 
+# function_scoped_fixture is only used in Python 3 compatible release of hypothesis
```
Comment (chaen) [inline src/DIRAC/Core/Utilities/test/Test_Encode.py:30]: Actually for that specific case, I think we could convert it to a module or scope fixture. It does not really need to change at each function. We could run all the diset test, then all the json, etc

**Comment by chrisburr** [comment]
Actually the context manager approach causes the tests to become flaky as hypothesis releases the performance varies significantly between otherwise identical runs so I've modified this to apply the setting to ignore the health check.

File: src/DIRAC/Core/Utilities/test/Test_Encode.py (hunk 4 above)
Comment (chrisburr) [reply src/DIRAC/Core/Utilities/test/Test_Encode.py:30]: I don't think we can as the environment variables need to be different between different functions in the same module as they're parametrised?

Comment (chaen) [reply src/DIRAC/Core/Utilities/test/Test_Encode.py:30]: Indeed not, I thought we could run it differently

**Comment by fstagni** [comment]
@chaen review OK?

**Comment by chaen** [comment]
yes, no choice I am afraid

**Review by chaen — APPROVED** [review]

**Pull request merged**